File: src/frontend/src/services/api/types/audit.ts
// Audit-log DTOs shared by the user-facing job list (services/api/conversion.ts's
// myJobs) and the admin audit log/runs (services/api/adminAudit.ts) — both
// endpoints return the same row shape, just filtered differently server-side.

export interface AuditEntry {
  id: number;
  ts: string | null;
  user_sub: string | null;
  user_email: string | null;
  user_display_name: string | null;
  scope_kind: string;
  scope_id: string | null;
  action: string;
  key: string | null;
  target_format: string | null;
  status: string | null;
  error: string | null;
  duration_ms: number | null;
  traceback: string | null;
  cpu_user_ms: number | null;
  cpu_sys_ms: number | null;
  peak_rss_kb: number | null;
  read_bytes: number | null;
  write_bytes: number | null;
  profile_key: string | null;
  job_id: string | null;
  // M5b: per-row issue-bot sync state for failed user conversions.
  // NULL until the bot has touched the row. Audit-run-attached
  // failures (audit_run_id IS NOT NULL) are processed via the
  // parent run's pass and leave this column NULL by design.
  audit_run_id: string | null;
  issue_bot_status: string | null;
  issue_bot_synced_at: string | null;
  issue_bot_last_error: string | null;
  // Stable per-device id (from client_metrics) — distinguishes view/render
  // audit logs by device (e.g. phone vs desktop). Null for server-side rows.
  device_id: string | null;
  // The worker image that processed a convert row; links to its package manifest.
  worker_image_tag: string | null;
  // Conversion engine + effective toggles for a convert row (which tessellator
  // actually ran, incl. an adacpp→occ-builtin fallback, + the options used).
  convert_meta: ConvertMeta | null;
}

export interface ConvertMeta {
  tessellator?: string;
  step_glb_pipeline?: string;
  glb_compression?: string;
  stream_workers?: string | number | null;
  // Wall-clock split of the recorded duration: the conversion proper vs the
  // GLB-compression (meshopt) post-step, so compression cost isn't mistaken
  // for a slower conversion.
  convert_ms?: number | null;
  compress_ms?: number | null;
  // The pod's CPU allotment (cgroup quota) at conversion time, so the metrics chart can render CPU
  // as % utilization across all cores rather than a cumulative ramp.
  cpu_cores?: number | null;
  options?: Record<string, string>;
  // adacpp [STEPPROF-JSON] pipeline summaries, parsed from the captured child log when
  // the profile_conversions toggle was on — one entry per instrumented C++ pipeline run.
  cpp_profile?: CppProfile[];
}

export interface CppProfilePhase {
  name: string;
  ms: number;
  rss_mb: number;
}

export interface CppProfileThread {
  tid: number;
  solids: number;
  busy_ms: number;
}

export interface CppProfile {
  label: string;
  wall_ms: number;
  peak_rss_mb: number;
  cpu_s?: number;
  parallelism?: number;
  vctx?: number;
  nvctx?: number;
  disk_read_mb?: number;
  majflt?: number;
  solids?: number;
  tris?: number;
  max_tris_solid?: number;
  phases: CppProfilePhase[];
  notes?: Record<string, number>;
  threads?: CppProfileThread[];
}

export interface AuditFilters {
  user_sub?: string;
  scope_kind?: string;
  scope_id?: string;
  action?: string;
  /** Conversion target format (glb / ifc / step / …). */
  target?: string;
  /** Job state (queued / running / done / error). */
  status?: string;
  /** Case-insensitive substring filter on the source filepath/filename. */
  key?: string;
  /** Lower bound on ``ts``: a duration the SERVER resolves ("6h", "30d"), or an
   * ISO-8601 instant for a custom range. Relative forms are deliberately not
   * resolved here — a clock a few minutes fast would silently empty a
   * "last 5 minutes" view. */
  since?: string;
  /** Upper bound on ``ts``, ISO-8601. Only set for a custom range. */
  until?: string;
  before_id?: number;
  limit?: number;
}

/** Aggregate counts behind the Audit tab's Overview. */

/** Queue pressure right now. Ages are of jobs still WAITING — `ts` is the
 * enqueue time, so a queued row's age is its wait so far. Jobs that already ran
 * are absent on purpose: nothing records when a worker picked one up, so a
 * historical wait would have to be invented. */
export interface AuditCongestion {
  queued: number;
  running: number;
  /** Null when nothing is queued — distinct from 0, which would mean
   * "served instantly". */
  oldest_wait_s: number | null;
  mean_wait_s: number | null;
  median_wait_s: number | null;
  /** How many rows carry a recorded start. Rows predating the started_at
   * migration do not, so this travels with the numbers — a median over three
   * rows deserves less trust than one over three thousand. */
  served: number;
  served_mean_wait_s: number | null;
  served_median_wait_s: number | null;
  served_p95_wait_s: number | null;
  served_max_wait_s: number | null;
}

export interface AuditSummary {
  total: number;
  congestion: AuditCongestion;
  /** Always carries the four states the queue writes, zero-filled. */
  by_status: Record<string, number>;
  by_target: { target: string; counts: Record<string, number>; total: number }[];
  top_errors: { error: string; count: number }[];
}

File: src/ada/comms/rest/worker/audit.py
"""Audit-row bookkeeping: final outcome patching (DB or over the API), conversion provenance and
the adacpp profiler summaries parsed out of the captured child log.
"""

from __future__ import annotations

import asyncio
import pathlib
import time

import asyncpg

from ada.config import logger

from .. import db as db_module
from .. import failure_capture
from ..queue import Job
from . import state
from .source_nodes import _rest_source_nodes_config


async def _report_job_status_over_api(job_id: str, payload: dict) -> bool:
    """Tell the API what a job is doing, for a worker with no database pool.

    WHY: both audit hops -- the running mark and the final outcome -- are gated on
    a pool, so on a pool-less worker the audit row stays `queued` for ever while
    the job runs, finishes and is swept. The queue record is accurate the whole
    time, so the conversion toast follows along and the Audit tab shows every job
    on that pool as permanently pending. An operator then cannot tell a healthy
    pool from a broken one, and anything reasoning over non-terminal rows blocks on
    jobs that finished minutes ago.

    The same argument, and the same mechanism, as the source-node REST recorder: a
    worker that already reaches this API should not need Postgres credentials to
    say what it just did.

    BEST EFFORT, ALWAYS. Returns whether it landed, and never raises: a report is
    commentary on work that has already happened, and failing a finished job
    because its commentary did not arrive would be strictly worse than the gap this
    closes. Blocking HTTP on the event loop's executor, because this is called from
    async code and the recorder idiom here is urllib.
    """
    cfg = _rest_source_nodes_config()
    if cfg is None:
        return False
    base, token = cfg

    def _post() -> bool:
        import json as _json
        import urllib.error
        import urllib.parse
        import urllib.request

        url = f"{base}/api/jobs/{urllib.parse.quote(job_id, safe='')}/status"
        body = _json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(url, data=body, method="POST")
        req.add_header("Authorization", f"Bearer {token}")
        req.add_header("Content-Type", "application/json")
        req.add_header("Accept", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=20.0) as resp:
                resp.read()
            return True
        except urllib.error.HTTPError as exc:
            # THE BODY IS READ AND LOGGED, not just the status. A bare code is how
            # two separate failures today each cost an hour: the server says what
            # went wrong in the body, and throwing it away leaves the reader
            # inferring from a number. Read defensively -- a body that cannot be
            # read must not replace the error with a different one.
            detail = ""
            try:
                detail = exc.read().decode("utf-8", "replace")[:500]
            except Exception:  # noqa: BLE001
                pass
            # Logged at warning rather than retried: a 4xx will be just as wrong
            # next time.
            #
            # 404/405 IS NAMED, because the bare status is actively misleading. A
            # viewer that predates this route has no handler for the path, so the
            # SPA's catch-all answers the GET shape and POST comes back 405 -- a
            # status that reads as "the API refused this" when it means "this API
            # does not have the feature yet". An afternoon went into a 405 that
            # meant something equally structural, so this one says what to do.
            if exc.code in (404, 405):
                logger.warning(
                    "worker: this viewer has no POST /api/jobs/{id}/status route (%s), so job "
                    "outcomes stay `queued` in the audit log until it is updated. The jobs "
                    "themselves are unaffected.",
                    exc.code,
                )
            else:
                # A 5xx here is the API failing to record, which is worth the body:
                # it is the only place the reason exists.
                logger.warning(
                    "worker: audit report for job %s refused (%s)%s",
                    job_id,
                    exc.code,
                    f": {detail}" if detail else "",
                )
            return False
        except Exception as exc:  # noqa: BLE001 - commentary must not sink a job
            logger.warning("worker: audit report for job %s did not reach the API: %s", job_id, exc)
            return False

    try:
        return await asyncio.get_running_loop().run_in_executor(None, _post)
    except Exception:  # noqa: BLE001
        logger.exception("worker: audit report for job %s could not be dispatched", job_id)
        return False


async def _audit_done(
    db_pool: asyncpg.Pool | None,
    job_id: str,
    status: str,
    error: str | None,
    started_at: float,
    traceback: str | None = None,
    metrics: dict | None = None,
) -> None:
    """Patch the audit_log row for this job with its final outcome.
    Best-effort: a DB hiccup must never break job processing."""
    if db_pool is None:
        # A worker without a pool reports over the API instead. Without this the
        # row never leaves `queued`, so every job this pool ever ran reads as
        # pending -- see _report_job_status_over_api.
        metrics = metrics or {}
        await _report_job_status_over_api(
            job_id,
            {
                "status": status,
                "error": error,
                "traceback": traceback,
                "duration_ms": int((time.time() - started_at) * 1000),
                "worker_image_tag": state._WORKER_IMAGE_TAG,
                **{
                    key: metrics[key]
                    for key in ("cpu_user_ms", "cpu_sys_ms", "peak_rss_kb", "read_bytes", "write_bytes")
                    if metrics.get(key) is not None
                },
            },
        )
        return
    metrics = metrics or {}
    # Preserve the input of a failed job while it still exists: the row outlives
    # the blob, and a user-scope source can be deleted at any time. Resolved from
    # the row rather than the call site so every error path is covered by this one
    # hook. Best-effort and deduplicated; None when disabled or ineligible.
    failure_key = None
    if status == "error" and state._WORKER_STORAGE is not None:
        failure_key = await failure_capture.capture_for_job(state._WORKER_STORAGE, db_pool, db_module, job_id)
    try:
        await db_module.update_audit_by_job(
            db_pool,
            job_id=job_id,
            status=status,
            error=error,
            duration_ms=int((time.monotonic() - started_at) * 1000),
            traceback=traceback,
            cpu_user_ms=metrics.get("cpu_user_ms"),
            cpu_sys_ms=metrics.get("cpu_sys_ms"),
            peak_rss_kb=metrics.get("peak_rss_kb"),
            read_bytes=metrics.get("read_bytes"),
            write_bytes=metrics.get("write_bytes"),
            profile_key=metrics.get("profile_key"),
            log_key=metrics.get("log_key"),
            worker_image_tag=state._WORKER_IMAGE_TAG,
            convert_meta=metrics.get("convert_meta"),
            failure_key=failure_key,
        )
    except Exception:
        logger.exception("worker: audit update failed for job %s", job_id)


def _convert_meta_for(job: "Job", env_overrides: dict | None) -> dict | None:
    """Provenance for a conversion's audit row: which tessellator/engine actually
    ran (resolved here the same way the convert subprocess resolves it — adacpp
    availability is identical in this shared env — so a libtess2→occ-builtin
    fallback is recorded accurately) plus the effective toggle options."""
    suffix = pathlib.PurePosixPath(job.source_key).suffix.lower()
    meta: dict = {}
    # The effective non-default toggles applied to the child (settings + per-job).
    if env_overrides:
        meta["options"] = dict(env_overrides)
    if job.target_format == "glb" and suffix in {".step", ".stp"}:
        try:
            from ada.comms.rest.converter import (
                _STEP_GLB_PIPELINE_ADACPP_NATIVE,
                _STEP_GLB_PIPELINE_OCC,
                _cad_config_for_pipeline,
                _resolve_step_glb_pipeline,
            )

            requested = _resolve_step_glb_pipeline((env_overrides or {}).get("ADAPY_STEP_GLB_PIPELINE"))
            meta["step_glb_pipeline"] = requested
            if requested == _STEP_GLB_PIPELINE_ADACPP_NATIVE:
                # Fully in-process C++ reader + tessellate + GLB writer — no CadBackend config, so
                # _cad_config_for_pipeline() is None for it (don't mislabel that as an occ fallback).
                meta["tessellator"] = "adacpp:native"
            elif requested == _STEP_GLB_PIPELINE_OCC:
                meta["tessellator"] = "occ-builtin"
            else:
                cfg = _cad_config_for_pipeline(requested)
                if cfg is not None:
                    meta["tessellator"] = cfg.path.value  # e.g. "adacpp:libtess2"
                else:
                    meta["tessellator"] = f"occ-builtin (fallback from {requested})"
            meta["glb_compression"] = (env_overrides or {}).get("ADA_GLB_COMPRESSION") or "meshopt"
            meta["stream_workers"] = (env_overrides or {}).get("ADA_STEP_STREAM_WORKERS")
        except Exception:
            logger.exception("worker: convert_meta tessellator resolution failed for %s", job.source_key)
    return meta or None


def _attach_cpp_profiles(convert_meta: dict | None, log_bytes: bytes | None) -> None:
    """Parse the adacpp pipeline profiler's machine-readable summaries out of the
    captured child output into ``convert_meta["cpp_profile"]``.

    When the ``profile_conversions`` toggle is on, the child runs with
    ``ADACPP_STEP_PROFILE=1`` and each instrumented C++ pipeline prints ONE
    ``[STEPPROF-JSON] {...}`` line at teardown (phase wall/RSS, VmHWM peak,
    per-solid stats, parallelism/IO pressure, per-thread utilisation). Attaching
    them to convert_meta puts the C++ side in the audit Metrics panel with the
    same visibility as the Python timings. No-op when profiling was off (no
    marker lines) or the log is empty."""
    if not isinstance(convert_meta, dict) or not log_bytes:
        return
    import json

    marker = b"[STEPPROF-JSON] "
    profiles: list[dict] = []
    for line in log_bytes.splitlines():
        i = line.find(marker)
        if i < 0:
            continue
        try:
            profiles.append(json.loads(line[i + len(marker) :].decode("utf-8", "replace")))
        except (ValueError, UnicodeDecodeError):
            continue  # a torn/interleaved line must not fail the job
    if profiles:
        convert_meta["cpp_profile"] = profiles

    # Per-conversion quality flags emitted by the child (subprocess_convert). Currently the
    # NGEOM(libtess2/adacpp)->OCC fallback tally — a conversion that silently completed on OCC
    # instead of the selected stream kernel. Surfaced as a cell flag in the audit grid.
    fb_marker = b"[TESSFALLBACK-JSON] "
    for line in log_bytes.splitlines():
        i = line.find(fb_marker)
        if i < 0:
            continue
        try:
            fb = json.loads(line[i + len(fb_marker) :].decode("utf-8", "replace"))
        except (ValueError, UnicodeDecodeError):
            continue
        if isinstance(fb, dict) and fb.get("count"):
            convert_meta["occ_fallback"] = fb  # {count, reasons, geoms}
        break

    mh_marker = b"[MESHHEALTH-JSON] "
    for line in log_bytes.splitlines():
        i = line.find(mh_marker)
        if i < 0:
            continue
        try:
            mh = json.loads(line[i + len(mh_marker) :].decode("utf-8", "replace"))
        except (ValueError, UnicodeDecodeError):
            continue
        if isinstance(mh, dict) and mh.get("distorted_tris"):
            convert_meta["mesh_flags"] = mh  # {n_tris, distorted_tris, distorted_frac}
        break

    # Triangle tally — total output triangles (+ primitive/solid counts when known). The primary
    # run-to-run regression signal: a tessellation-density change or a dropped solid moves n_tris.
    ts_marker = b"[TRISTATS-JSON] "
    for line in log_bytes.splitlines():
        i = line.find(ts_marker)
        if i < 0:
            continue
        try:
            ts = json.loads(line[i + len(ts_marker) :].decode("utf-8", "replace"))
        except (ValueError, UnicodeDecodeError):
            continue
        if isinstance(ts, dict) and ts.get("n_tris"):
            convert_meta["tri_stats"] = ts  # {n_tris, engine?, n_primitives?, n_solids?, ...}
        break

    # Geometry health — faces with a real trim boundary that tessellated to zero triangles (silently
    # dropped geometry, e.g. the SURFACE_OF_LINEAR_EXTRUSION drops). Flagged in the audit grid so this
    # class of bug is caught without visual inspection.
    gh_marker = b"[GEOMHEALTH-JSON] "
    for line in log_bytes.splitlines():
        i = line.find(gh_marker)
        if i < 0:
            continue
        try:
            gh = json.loads(line[i + len(gh_marker) :].decode("utf-8", "replace"))
        except (ValueError, UnicodeDecodeError):
            continue
        if isinstance(gh, dict) and gh.get("dropped_faces"):
            convert_meta["geom_health"] = gh  # {dropped_faces, total_faces}
        break
